### ISSUE TITLE
Attachments not indexed when imported via WP All Import

### ISSUE THREAD

**Issue by tamara-m** [10up/ElasticPress#2040]
**Describe the bug**
When importing a set of ~2,000 attachment images through a WP All Import import job, attachments don't get indexed. 
The images files are uploaded to the server, and all the data and metadata get saved to the posts and postmeta tables correctly and completely. 

But, the media library displays empty, as ElasticPress finds 0 hits for attachments.
Manually running `wp elasticpress index --post-type=attachment` after the import, correctly indexes all attachments and everything works as expected

**Steps to Reproduce**

Run a WP All Import job that involves importing images.

**Expected behavior**
Attachments newly created in the database via the import should get indexed in ElasticSearch.

**Request**
You might be interested in the compatibility with WP All Import, or you may not, I understand you cannot provide compatibility for everything and I am very thankful for the fantastic product that ElasticPress is - as is.

I was wondering if you could direct me to any resource so that I can either index all attachments programmatically once the import has completed, or one at a time, as they get added/updated in the db.

Thank you in advance for any direction,

**Comment by tamara-m** [comment]
This is not a bug, ElasticPress turns off indexing while importing by design, which I can understand.

We have a daily import job though and still needed an automation, so for everyone needing to index posts programmatically, the answer was using

`\ElasticPress\Indexables::factory()->get( 'post' )->index( $post_id, true );`

**Comment by bogdanarizancu** [comment]
@tamara-m I am also facing the issue of imported products using `WP All Import` not being indexed, what I found was that setting the `EP_SYNC_CHUNK_LIMIT` will trigger the sync. I have it set to zero, since we may have only one or two products on an import, so this way it will always trigger. The code related to this is documented [here](https://github.com/10up/ElasticPress/blob/906137030cfb9a7dfdde470b483e35c379d55ad7/includes/classes/SyncManager.php#L43-L49).

**Comment by ouch-digital** [comment]
Thanks @bogdanarizancu this helped me :)

**Comment by spithash** [comment]
I'm having the same issue when importing products, even deleted items do not get erased from the index automatically, is there a solution to this? I tried setting EP_SYNC_CHUNK_LIMIT to 0 without any luck. 

Can anyone suggest something? I'm stuck.

**Comment by tamara-m** [comment]
All I can think of is triggering the deletion form index programmatically, similar to how I triggered the indexing above, but hooked to the delete product action. Hope this helps

**Comment by spithash** [comment]
@tamara-m I mean, they both don't work, deleted and new imported items. I always have to sync by myself. I mean, I trigger it via cronjob every x hours or so via wp-cli but, there should be a better way, idk...

By the way, how (and where) do you use that line below?
`\ElasticPress\Indexables::factory()->get( 'post' )->index( $post_id, true );`

Thank you for the reply

**Comment by tamara-m** [comment]
I'm using that inside of a custom plugin I developed. You'd need to do the same or learn how to add php code snippets to your WordPress site.

**Comment by spithash** [comment]
Ah, gotcha :) Thanks...

**Comment by barthusz** [comment]
For anyone who is still struggling with this, here's a little more explanation about how to fix this:

You need to use the pmxi_saved_post hook which is described here: https://www.wpallimport.com/documentation/action-reference/

And to show you how I fixed it; I only needed to add this script in the functions.php of my theme file (or you can put in in a custom plugin):

function elastic_update($id) {     
    \ElasticPress\Indexables::factory()->get( 'post' )->index( $id, true );
}

add_action('pmxi_saved_post', 'elastic_update', 10, 1);

Basically it updates the index automatically with every post import. Hope this helps!
